spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/subscriber/PubSubSubscriberTemplate.java

```diff
@@ -315,6 +281,67 @@ public void onSuccess(Empty empty) {
 		return this.subscriberStub.modifyAckDeadlineCallable().futureCall(modifyAckDeadlineRequest);
 	}
 
+	/**
+	 * Perform Pub/Sub operations (ack/nack/modifyAckDeadline) in per-subscription batches.
+	 * <p>The returned {@link ListenableFuture} will complete when either all batches completes successfully or when at
+	 * least one fails.</p>
+	 * <p>
+	 * In case of multiple batch failures, which exception exception will be in the final {@link ListenableFuture} is
+	 * non-deterministic.
+	 * </p>
+	 * @param acknowledgeablePubsubMessages messages, could be from different subscriptions.
+	 * @param asyncOperation Pub/Sub operation to perform.
+	 * @return ListenableFuture indicating overall success or failure.
+	 */
+	private ListenableFuture<Void> doBatchedAsyncOperation(
+			Collection<AcknowledgeablePubsubMessage> acknowledgeablePubsubMessages,
+			BiFunction<String, List<String>, ApiFuture<Empty>> asyncOperation) {
+
+		Map<ProjectSubscriptionName, List<String>> groupedMessages
+				= acknowledgeablePubsubMessages.stream()
+				.collect(
+						Collectors.groupingBy(
+								AcknowledgeablePubsubMessage::getProjectSubscriptionName,
+								Collectors.mapping(AcknowledgeablePubsubMessage::getAckId, Collectors.toList())));
+
+		Assert.state(groupedMessages.keySet().stream().map(ProjectSubscriptionName::getProject).distinct().count() == 1,
+				"The project id of all messages must match.");
+
+		SettableListenableFuture<Void> settableListenableFuture	= new SettableListenableFuture<>();
+		ExecutorService cachedThreadPoolExecutorService = Executors.newCachedThreadPool();
```

Comment:
I'm not sure we should use a cached thread pool here, especially since we're not allowing it to be configured. I would just go with a direct executor. WDYT?

Reply:
Direct is fine; it's consistent with previous behavior where the executor selection was delegated to ApiFutures.addCallback(). I just did not want to keep the deprecated version of addCallback with unspecified executor.

Reply:
That's correct. Only my concern that you create a new `Executors.newCachedThreadPool()` for each method call.
This is not acceptable. The thread executor should be global one for the whole instance and should be destroyed in the end of component lifecycle.
Also we need to provide a setter to allow to inject an externally managed executor.

That's nothing more.

Reply:
@artembilan Did you have in mind a separate thread pool per each PubSubSubscriberTemplate instance?

Reply:
Well, typically we have only a single `PubSubSubscriberTemplate` in our microservice application, so that's OK to have some default internal `ThreadPoolTaskExecutor` in the `PubSubSubscriberTemplate`.

With an extra setter the user has option to provide a managed executor from the application context.